include/boost/math/distributions/dirichlet.hpp

```diff
@@ -0,0 +1,405 @@
+  } // dirichlet_distribution constructor.
+
+  // Accessor functions:
+  VectorType Concentration() const
+  {
+    return concentration;
+  }
+
+  size_t Order() const
+  {
+    return concentration.size();
+  }
+
+  static VectorType find_concentration(
+      VectorType mean,     // Expected value of mean.
+      VectorType variance) // Expected value of variance.
+  {
+    assert(("Dimensions of mean and variance must be same!", mean.size() == variance.size()));
+    static const char *function = "boost::math::dirichlet_distribution<%1%>::find_concentration";
+    RealType result = 0; // of error checks.
+    if (!(dirichlet_detail::check_mean(function, mean, &result, Policy()) && dirichlet_detail::check_variance(function, variance, &result, Policy())))
+    {
+      return result;
+    }
+    VectorType c;
+    for (size_t i = 0; i < mean.size(); ++i)
+    {
+      c.push_back(mean[i] * (((mean[i] * (1 - mean[i])) / variance[i]) - 1));
+    }
+    return c;
+  } // RealType find_concentration
+
+  // TODO
+  // static VectorType find_concentration(
+  //     VectorType x,           //  x.
+  //     VectorType probability) // cdf
+  // {
+  //   assert(("", x.size() == probability.size()));
+  //   static const char *function = "boost::math::dirichlet_distribution<%1%>::find_conentration";
+  //   RealType result = 0; // of error checks.
+  //   if (!(dirichlet_detail::check_prob(function, probability, &result, Policy()) && dirichlet_detail::check_x(function, x, &result, Policy())))
+  //   {
+  //     return result;
+  //   }
+  //   return ;
+  // } // RealType find_concentration(x, probability)
+
+private:
+  VectorType concentration; // https://en.wikipedia.org/wiki/Concentration_parameter.
+  RealType sum_concentration;
+}; // template <class RealType, class Policy> class dirichlet_distribution
+
+
+template <class VectorType, class RealType, class Policy>
+inline const std::pair<RealType, RealType> range(const dirichlet_distribution<VectorType, RealType, Policy> & /* dist */)
+{ // Range of permissible values for random variable x.
+  using boost::math::tools::max_value;
+  return std::pair<RealType, RealType>(static_cast<RealType>(0), static_cast<RealType>(1));
+}
+
+
+template <class VectorType, class RealType, class Policy>
+inline const std::pair<RealType, RealType> support(const dirichlet_distribution<VectorType, RealType, Policy> & /* dist */)
+{ // Range of supported values for random variable x.
+  // This is range where cdf rises from 0 to 1, and outside it, the pdf is zero.
+  return std::pair<RealType, RealType>(static_cast<RealType>(0), static_cast<RealType>(1));
+}
+
+
+template <class VectorType, class RealType, class Policy>
+inline VectorType mean(const dirichlet_distribution<VectorType, RealType, Policy> &dist)
+{ // Mean of dirichlet distribution = c[i]/sum(c).
+  VectorType m;
+  for (size_t i = 0; i < dist.Order(); ++i)
+  {
+    m.push_back(dist.concentration[i] / dist.sum_concentration);
```

Comment:
Repeatedly calling `push_back` is bad; get it to the correct size in line 255 and then just fill it up.